File: py/application/games/game.py
from .constants import GamePhase


class Game():

  def __init__(self, gameSettings=None, gameState=None, gamePhase=None):
    self._gameSettings = gameSettings
    self._gameState = gameState
    self._gamePhase = gamePhase

  @property
  def gameSettings(self):
    return self._gameSettings

  @property
  def gameState(self):
    return self._gameState

  @property
  def gamePhase(self):
    return self._gamePhase

  @classmethod
  def getDefaultPlayerConfig(cls):
    return [
      {'playerType': 'human', 'owner': None, 'name': 'Player 1', 'style': 'A'},
      {'playerType': 'human', 'owner': None, 'name': 'Player 2', 'style': 'B'},
    ]

  @classmethod
  def getSettingsConfig(cls):
    return []

  @classmethod
  def getInitialGameState(cls, gameSettings=None):
    return {}

  def start(self):
    self._gamePhase = GamePhase.PLAYING.value
    self.nextPlayerTurn()

  def nextPlayerTurn(self):
    # By default, assumes two players alternating
    if 'gameEnd' in self._gameState and self._gameState['gameEnd']:
      self._gameState['activePlayerIndex'] = None
    elif ('activePlayerIndex' not in self._gameState or
          self._gameState['activePlayerIndex'] is None):
      self._gameState['activePlayerIndex'] = 0
    else:
      numPlayers = len(self._gameSettings['players'])
      self._gameState['activePlayerIndex'] = (
          self._gameState['activePlayerIndex'] + 1) % numPlayers

  def checkGameEndCondition(self):
    self._gameState['gameEnd'] = self.gameEndCondition()

  def gameEndCondition(self):
    return None

  def action(self, action):
    return None

File: node/stuff/Game.ts
import {BaseGameSettings} from './BaseGameSettings'
import {SettingsConfig} from './SettingsConfig'
import {BaseGameState} from './BaseGameState'

export type UnknownGame = Game<BaseGameState, unknown, BaseGameSettings>

export abstract class Game<TGameState extends BaseGameState, TGameAction, TGameSettings extends BaseGameSettings> {
  gameState: TGameState
  gamePhase: number = 0
  gameSettings: TGameSettings

  constructor(
    gameState: TGameState | null = null,
    gameSettings: TGameSettings | null = null,
    gamePhase: number = 1  // PRE-GAME
  ) {

    if (gameState && gameSettings) {
      this.gameSettings = gameSettings
      this.gameState = gameState
      this.gamePhase = gamePhase
      return
    }

    const playerConfig = this.defaultPlayerConfig()
    const settings = this.buildDefaultSettings()
    settings.players = playerConfig

    this.gameState = this.initialGameState(settings)
    this.gameSettings = settings
    return
  }

  defaultPlayerConfig() {
    return [
      {'playerType': 'human', 'owner': null, 'name': 'Player 1', 'style': 'A'},
      {'playerType': 'human', 'owner': null, 'name': 'Player 2', 'style': 'B'},
    ]
  }

  buildDefaultSettings() {
    let settings: any = {}
    this.settingsConfig().forEach(setting => {
      settings[setting.canonicalName] = setting.defaultValue
    })
    return settings
  }

  abstract action(action: TGameAction): boolean
  abstract canonicalName: string

  checkGameEndCondition() {
    const gameEnd = this.gameEndCondition()
    if (!!gameEnd) {
      this.gamePhase = 3 // GAME-OVER
    }

    this.gameState.gameEnd = this.gameEndCondition()
  }

  start() {
    this.gamePhase = 2 // IN-PROGRESS
    this.nextPlayerTurn()
  }

  nextPlayerTurn() {
    // By default, assumes two players alternating
    if (this.gameState.gameEnd) {
      this.gameState.activePlayerIndex = null
    } else if (this.gameState.activePlayerIndex === null) {
      this.gameState.activePlayerIndex = 0
    } else {
      const numPlayers = this.gameSettings.players.length
      this.gameState.activePlayerIndex = (this.gameState.activePlayerIndex + 1) % numPlayers
    }
  }

  abstract gameEndCondition(): any;

  abstract settingsConfig(): SettingsConfig[]

  abstract initialGameState(settings: TGameSettings): TGameState
}
